### PR TITLE
Optimise has many notifications

### PR DESCRIPTION
**Pull request by BryanCrotaz**
Diff many arrays properly to find the smallest contiguous block describing the change

Particularly important when polling server for changes with a repeating model.reload() or relationship.update()

### CODE REVIEW

**Comment by BryanCrotaz** [comment]
Good point.

File: addon/-private/system/many-array.js [hunk 1]
```diff
@@ -145,25 +146,31 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
...
-      //TODO Figure out to notify only on additions and maybe only if unloaded
-      this.relationship.notifyHasManyChanged();
+    // diff to find changes
+    const diff = diffArray(this.currentState, toSet);
+
+    if (diff.firstChangeIndex !== null) { // it's null if no change found
+      // we found a change
+      this.arrayContentWillChange(diff.firstChangeIndex, diff.removedCount, diff.addedCount);
+      // It’s possible the parent side of the relationship may have been unloaded by this point
+      if (_objectIsAlive(this)) {
```
Comment (stefanpenner) [inline addon/-private/system/many-array.js]: if we are not alive, i think we should just skip the whole function, exiting at line 145.

File: addon/-private/system/diff-array.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  @for DS
+  @param {Array} oldArray the old array
+  @param {Array} newArray the new array
+  @return {hash} {
+      firstChangeIndex: <integer>,  // null if no change
+      addedCount: <integer>,        // 0 if no change
+      removedCount: <integer>       // 0 if no change
+    }
+*/
+const diffArray = function (oldArray, newArray) {
```
Comment (stefanpenner) [inline addon/-private/system/diff-array.js]: do you mind if we do `export default function diffArray`

* This lets readers easily see (without scanning the full file) that this method is the default export

File: addon/-private/system/many-array.js [hunk 3]
```diff
@@ -141,29 +142,36 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
...
+    // It’s possible the parent side of the relationship may have been unloaded by this point
+    if (!_objectIsAlive(this)) {
+      return;
+    }
     let toSet = this.canonicalState;
 
     //a hack for not removing new records
     //TODO remove once we have proper diffing
-    let newRecords = this.currentState.filter(
+    const newRecords = this.currentState.filter(
```
Comment (stefanpenner) [inline addon/-private/system/many-array.js]: lets leave `let` here

File: addon/-private/system/many-array.js [hunk 4]
```diff
@@ -141,29 +142,36 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
...
-      this.set('length', toSet.length);
-    }
-    this.currentState = toSet;
-    this.arrayContentDidChange(0, oldLength, this.length);
 
-    if (isInitialized) {
-      //TODO Figure out to notify only on additions and maybe only if unloaded
-      this.relationship.notifyHasManyChanged();
+    // diff to find changes
+    const diff = diffArray(this.currentState, toSet);
```
Comment (stefanpenner) [inline addon/-private/system/many-array.js]: `let`

File: addon/-private/system/many-array.js [hunk 5]
```diff
@@ -289,12 +297,11 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
...
   createRecord(hash) {
-    let store = get(this, 'store');
-    let type = get(this, 'type');
-    let record;
+    const store = get(this, 'store');
+    const type = get(this, 'type');
 
     assert(`You cannot add '${type.modelName}' records to this polymorphic relationship.`, !get(this, 'isPolymorphic'));
-    record = store.createRecord(type.modelName, hash);
+    const record = store.createRecord(type.modelName, hash);
```
Comment (stefanpenner) [inline addon/-private/system/many-array.js]: `let`

**Comment by stefanpenner** [comment]
eslint looks unhappy:

```
 18) [PhantomJS 2.1] ESLint - modules/ember-data/-private/system/diff-array.js: should pass ESLint
     modules/ember-data/-private/system/diff-array.js should pass ESLint.
58:2  - Unnecessary semicolon. (no-extra-semi)
     expected: true
       actual: false
     http://localhost:7357/assets/dummy.js:159:14
     runTest@http://localhost:7357/assets/test-support.js:849:32
     run@http://localhost:7357/assets/test-support.js:834:11
     http://localhost:7357/assets/test-support.js:976:14
     process@http://localhost:7357/assets/test-support.js:635:24
     begin@http://localhost:7357/assets/test-support.js:617:9
     http://localhost:7357/assets/test-support.js:677:9
```

**Comment by stefanpenner** [comment]
🎉 🎉 🎉

**Comment by BryanCrotaz** [comment]
:) We got there eventually

**Comment by stefanpenner** [comment]
between this and https://github.com/emberjs/data/pull/4842 hasMany relationships have gotten some good TLC this week.

**Comment by BryanCrotaz** [comment]
I just found another 10x improvement... more to come

**Comment by stefanpenner** [comment]
@BryanCrotaz nice. We are also working on more perf things. Sounds like ember-data will be quite different perf wise soon. 

Can't wait to see what your working on!

**Comment by BryanCrotaz** [comment]
@stefanpenner this is what we're working on...

**Pull request merged**